cub/cub/device/device_for.cuh

```diff
@@ -0,0 +1,817 @@
+  {
+    // Dereferencing `thrust::device_vector<T>` iterators returns a `thrust::device_reference<T>`
+    // instead of `T`. Since user-provided operator expects `T` as an argument, we need to unwrap.
+    op(THRUST_NS_QUALIFIER::raw_reference_cast(*(input + i)));
+  }
+};
+
+/**
+ * `op_wrapper_vectorized_t` turns bulk into a for-each-copy operation.
+ * `op_wrapper_vectorized_t` is similar to `op_wrapper_t` but does not provide any guarantees about
+ * address of the input parameter. `OpT` might be given a copy of the value or an actual reference
+ * to the input iterator value (depending on the alignment of input iterator)
+ */
+template <class OffsetT, class OpT, class T>
+struct op_wrapper_vectorized_t
+{
+  const T* input; // Raw pointer to the input data
+  OpT op; // User-provided operator
+  OffsetT partially_filled_vector_id; // Index of the vector that doesn't have all elements
+  OffsetT num_items; // Total number of non-vectorized items
+
+  // TODO Can be extracted into tuning
+  constexpr static int vec_size = 4;
+
+  // Type of the vector that is used to load the input data
+  using vector_t = typename CubVector<T, vec_size>::Type;
+
+  _CCCL_DEVICE _CCCL_FORCEINLINE void operator()(OffsetT i)
+  {
+    // Surrounding `Bulk` call doesn't invoke this operator on invalid indices, so we don't need to
+    // check for out-of-bounds access here.
+    if (i != partially_filled_vector_id)
+    { // Case of fully filled vector
+      const vector_t vec = *reinterpret_cast<const vector_t*>(input + vec_size * i);
+
+#pragma unroll
+      for (int j = 0; j < vec_size; j++)
+      {
+        op(*(reinterpret_cast<const T*>(&vec) + j));
+      }
+    }
+    else
+    { // Case of partially filled vector
+      for (OffsetT j = i * vec_size; j < num_items; j++)
+      {
+        op(input[j]);
+      }
+    }
+  }
+};
+
+} // namespace for_each
+} // namespace detail
+
+struct DeviceFor
+{
+private:
+  /**
+   * Checks if the pointer is aligned to the given vector type
+   */
+  template <class VectorT, class T>
+  CUB_RUNTIME_FUNCTION static bool is_aligned(const T* ptr)
+  {
+    return (reinterpret_cast<std::size_t>(ptr) & (sizeof(VectorT) - 1)) == 0;
+  }
+
+  template <class InputIteratorT, class OffsetT, class OpT>
+  CUB_RUNTIME_FUNCTION static cudaError_t for_each_n(
+    InputIteratorT first,
+    OffsetT num_items,
+    OpT op,
+    cudaStream_t stream,
+    ::cuda::std::false_type /* do_not_vectorize */)
+  {
+    using wrapped_op_t = detail::for_each::op_wrapper_t<OffsetT, OpT, InputIteratorT>;
+    return detail::for_each::dispatch_t<OffsetT, wrapped_op_t>::dispatch(num_items, wrapped_op_t{first, op}, stream);
+  }
+
+  template <class InputIteratorT, class OffsetT, class OpT>
+  CUB_RUNTIME_FUNCTION static cudaError_t for_each_n(
+    InputIteratorT first, OffsetT num_items, OpT op, cudaStream_t stream, ::cuda::std::true_type /* vectorize */)
+  {
+    auto unwrapped_first = THRUST_NS_QUALIFIER::raw_pointer_cast(&*first);
```

Comment:
Should this use `cuda::std::addressof`?

Reply:
What happens here is:
```cpp
  thrust::device_vector<int> vec(10);
  thrust::device_vector<int>::iterator begin = vec.begin();
  thrust::device_reference<int> thrust_ref = *begin;
  thrust::device_ptr<int> thrust_ptr = &thrust_ref;
  int* actual_ptr = thrust::raw_pointer_cast(thrust_ptr);
```
There's actual `operator&` that we need to invoke as opposed to taking address of `thrust::device_reference`.